File: lib/core/domain/usecases/get_asset.dart
//
// Copyright (c) 2020 Nathan Fiedler
//
import 'package:equatable/equatable.dart';
import 'package:meta/meta.dart';
import 'package:oxidized/oxidized.dart';
import 'package:tanuki/core/domain/entities/asset.dart';
import 'package:tanuki/core/domain/repositories/entity_repository.dart';
import 'package:tanuki/core/domain/usecases/usecase.dart';
import 'package:tanuki/core/error/failures.dart';

class GetAsset implements UseCase<Asset, Params> {
  final EntityRepository repository;

  GetAsset(this.repository);

  @override
  Future<Result<Asset, Failure>> call(Params params) async {
    return await repository.getAsset(
      params.assetId,
    );
  }
}

class Params extends Equatable {
  final String assetId;

  Params({
    @required this.assetId,
  });

  @override
  List<Object> get props => [assetId];
}

File: server/domain/usecases/get-asset.ts
//
// Copyright (c) 2025 Nathan Fiedler
//
import assert from 'node:assert';
import { Asset } from 'tanuki/server/domain/entities/asset.ts';
import { type RecordRepository } from 'tanuki/server/domain/repositories/record-repository.ts';

export default ({
  recordRepository
}: {
  recordRepository: RecordRepository;
}) => {
  assert.ok(recordRepository, 'record repository must be defined');
  /**
   * Retrieve an asset entity from the record repository.
   *
   * @returns asset entity, or null if not found.
   */
  return async (assetId: string): Promise<Asset | null> => {
    return await recordRepository.getAssetById(assetId);
  };
};
